ffi/src/scan.rs

```diff
@@ -83,27 +93,28 @@ unsafe fn get_raw_arrow_data_impl(
     Ok(Box::leak(ret_data))
 }
 
-impl BoxHandle for Scan {}
+#[handle_descriptor(target=Scan, mutable=true, sized=true)]
+pub struct ScanHandle;
```

Comment:
Rescuing https://github.com/delta-incubator/delta-kernel-rs/pull/182#discussion_r1603862042
> Actually it's not shared I guess, so either `MutScan` or `ExclusiveScan`?

Yeah scan is a weird one (duckdb was hitting concurrency bugs here). It has to be mutable, because consuming from the scan modifies its internal state. I think we fixed the bug by adding a mutex internally, but technically duckdb should probably be doing the mutex work to ensure there's only one mutable reference possible at a time??

Reply:
Thinking more about this -- we should be doing an `Arc<Mutex<T>>` for this case, instead of locking internally inside the iterator. Right now it may "work" but we're violating the mutable borrow invariant. Unfortunately, the existing `Mutex` seems needed in order for the result of `new_async_read_iterator` to be `Send`:
```
error[E0277]: `std::sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>` cannot be shared between threads safely
   --> kernel/src/engine/default/file_stream.rs:151:12
    |
151 |         Ok(Box::new(it))
    |            ^^^^^^^^^^^^ `std::sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>` cannot be shared between threads safely
    |
    = help: within `std::iter::Map<std::sync::mpsc::IntoIter<std::result::Result<arrow_array::RecordBatch, error::Error>>, {closure@kernel/src/engine/default/file_stream.rs:145:18: 145:23}>`, the trait `std::marker::Sync` is not implemented for `std:\
:sync::mpsc::Receiver<std::result::Result<arrow_array::RecordBatch, error::Error>>`
note: required because it appears within the type `IntoIter<Result<RecordBatch, Error>>`
   --> /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/std/src/sync/mpsc/mod.rs:302:12
note: required because it appears within the type `Map<IntoIter<Result<RecordBatch, Error>>, {closure@file_stream.rs:145:18}>`
   --> /rustc/79e9716c980570bfd1f666e3b16ac583f0168962/library/core/src/iter/adapters/map.rs:61:12
    = note: required for the cast from `Box<Map<IntoIter<Result<RecordBatch, Error>>, {closure@file_stream.rs:145:18}>>` to `Box<dyn Iterator<Item = Result<Box<dyn EngineData>, Error>> + Send + Sync>`
```